src/backends/policy_enforcement/souffle/value.h

```diff
@@ -0,0 +1,124 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#ifndef SRC_BACKENDS_POLICY_ENFORCEMENT_SOUFFLE_VALUE_H_
+#define SRC_BACKENDS_POLICY_ENFORCEMENT_SOUFFLE_VALUE_H_
+
+#include <string>
+#include <tuple>
+#include <variant>
+
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/string_view.h"
+
+namespace raksha::backends::policy_enforcement::souffle {
+
+// The common supertype of any "fully fledged" Souffle value type. Anything
+// descending form `Value` should be able to be used in a `.type` declaration.
+// By this logic, `number`, `symbol`, any record, and any ADT are `Value`s, but
+// individual branches of an ADT are not.
+class Value {
+ public:
+  virtual std::string ToDatalogString() const = 0;
+};
+
+// This is the common supertype of any Souffle value type that can
+// participate in an ADT. This is all Values except Records. Values that can
+// participate in an ADT must have a virtual destructor.
+class AdtMemberValue : public Value {
```

Comment:
`AdtBranchArgument` seems like a better name?  Also, this class seems like a good candidate for `variant`?

Reply:
I've removed `AdtMemberValue` entirely, as I had incorrect assumptions as to what was allowed in an ADT branch.

On `std::variant`, my first attempt at representing an ADT actually had the ADT be a thin wrapper around a `std::variant` containing the branches of the ADT. The problem with that approach is that ADT branches may themselves contain the ADT to which they belong as an argument. Because `std::variant` is a tagged union, containing all of its elements as values, this can cause layout issues if they are handled directly as values. We can get around this by using pointers to the ADT branch options. But we've entered a new problem: the template for the ADT contains the full types of the branches, which contain the full type for the ADT, causing infinite expansion. We can get around that by erasing the type, having an erased branch type containing a vector rather than a tuple. I pushed on this for a while and ran into issues where GCC got very unhappy with me and eventually quit without fully understanding.

So to sum up: I agree that it might be handy to have ADTs be represented as variants, but fact that ADT values can live within ADT branches makes that very complicated, especially with the static-typing-through-template tricks I'm trying to pull.